remotes/docker/registry.go

```diff
@@ -0,0 +1,173 @@
+)
+
+// RegistryCapabilities represent the capabilities of the
+// registry. This also represents the set of operations
+// for which the registry endpoint may be trusted to perform.
+//
+// For example pushing is a capability which should only be
+// performed on an upstream source, not a mirror.
+// Resolving (the process of converting a name into a digest)
+// must be considered a trusted operation and only done by
+// a host which is trusted (or more preferably by secure process
+// which can prove the provenance of the mapping). A public
+// mirror should never be trusted to do a resolve action.
+//
+// | Registry Type    | Pull | Resolve | Push |
+// |------------------|------|---------|------|
+// | Public Registry  | yes  | yes     | yes  |
+// | Private Registry | yes  | yes     | yes  |
+// | Public Mirror    | yes  | no      | no   |
+// | Private Mirror   | yes  | yes     | no   |
+type RegistryCapabilities uint8
+
+const (
+	// RegistryCapPull represents the capability to fetch manifests
+	// and blobs by digest
+	RegistryCapPull RegistryCapabilities = 1 << iota
+
+	// RegistryCapResolve represents the capability to fetch manifests
+	// by name
+	RegistryCapResolve
+
+	// RegistryCapPush represents the capability to push blobs and
+	// manifests
+	RegistryCapPush
+
+	// Reserved for future capabilities (i.e. search, catalog, remove)
+)
+
+// RegistryHost represents a complete configuration for a registry
+// host, representing the capabilities, authorizations, connection
+// configuration, and location.
+type RegistryHost struct {
+	Client       *http.Client
+	Authorizer   Authorizer
+	Host         string
+	Scheme       string
+	Path         string
+	Capabilities RegistryCapabilities
+}
+
+// RegistryHosts fetches the registry hosts for a given namespace,
+// provided by the host component of an distribution image reference.
+type RegistryHosts func(string) ([]RegistryHost, error)
+
+// Registries joins multiple registry configuration functions, using the same
+// order as provided within the arguments. When an empty registry configuration
+// is returned with a nil error, the next function will be called.
+// NOTE: This function will not join configurations, as soon as a non-empty
+// configuration is returned from a configuration function, it will be returned
+// to the caller.
+func Registries(registries ...RegistryHosts) RegistryHosts {
+	return func(host string) ([]RegistryHost, error) {
+		for _, registry := range registries {
+			config, err := registry(host)
+			if err != nil {
+				return config, err
+			}
+			if len(config) > 0 {
+				return config, nil
+			}
+		}
+		return nil, nil
+	}
+}
+
+type registryOpts struct {
+	authorizer Authorizer
+	plainHTTP  bool
+	host       func(string) (string, error)
+	client     *http.Client
+}
+
+// RegistryOpt defines a registry default option
+type RegistryOpt func(*registryOpts)
+
+// WithPlainHTTP configures registries to use HTTP by default
+func WithPlainHTTP(opts *registryOpts) {
```

Comment:
I feel `WithInsecureHTTP` would be more descriptive

Reply:
Not a fan of using `Insecure` at this level. This part is only structuring a configuration rather than opining about how secure it is. The clients can make that determination. I'll make the comment more clear that it is using plaintext "http" scheme, `HTTP` used in this context I think is incorrect.